include/libpmemobj++/experimental/radix.hpp

```diff
@@ -0,0 +1,1813 @@
+		!std::is_same<T, inline_string>::value,
+	typename radix_tree<Key, Value>::template radix_tree_iterator<
+		IsConst>::pointer>::type
+	radix_tree<Key, Value>::radix_tree_iterator<IsConst>::operator->() const
+{
+	return &node->get_leaf()->data;
+}
+
+template <typename Key, typename Value>
+template <bool IsConst>
+typename radix_tree<Key, Value>::const_key_reference
+radix_tree<Key, Value>::radix_tree_iterator<IsConst>::key() const
+{
+	return node->get_leaf()->key();
+}
+
+template <typename Key, typename Value>
+template <bool IsConst>
+template <typename T>
+typename std::enable_if<!std::is_same<T, inline_string>::value,
+			typename radix_tree<Key, Value>::reference>::type
+radix_tree<Key, Value>::radix_tree_iterator<IsConst>::value() const
+{
+	return node->get_leaf()->data.second;
+}
+
+template <typename Key, typename Value>
+template <bool IsConst>
+template <typename T>
+typename std::enable_if<std::is_same<T, inline_string>::value,
+			typename radix_tree<Key, Value>::reference>::type
+radix_tree<Key, Value>::radix_tree_iterator<IsConst>::value() const
+{
+	return radix_tree::reference(node);
+}
+
+template <typename Key, typename Value>
+template <bool IsConst>
+typename radix_tree<Key, Value>::template radix_tree_iterator<IsConst>
+radix_tree<Key, Value>::radix_tree_iterator<IsConst>::operator++()
+{
+	assert(node);
+
+	/* node is root, there is no other leaf in the tree */
+	if (!node->get_leaf()->parent)
+		node = nullptr;
+	else
+		node = const_cast<node_ptr>(
+			next_leaf(typename node::forward_iterator(
+				node, node->get_leaf()->parent.get_node())));
+
+	return *this;
+}
+
+template <typename Key, typename Value>
+template <bool IsConst>
+typename radix_tree<Key, Value>::template radix_tree_iterator<IsConst>
+radix_tree<Key, Value>::radix_tree_iterator<IsConst>::operator--()
+{
+	assert(node);
+
+	/* node is root, there is no other leaf in the tree */
+	if (!node->get_leaf()->parent)
+		node = nullptr;
+	else
+		node = const_cast<node_ptr>(
+			next_leaf(typename node::reverse_iterator(
+				node, node->get_leaf()->parent.get_node())));
+
+	return *this;
+}
+
+template <typename Key, typename Value>
+template <bool IsConst>
+bool
+radix_tree<Key, Value>::radix_tree_iterator<IsConst>::operator!=(
+	const radix_tree_iterator<IsConst> &rhs) const
+{
+	return node != rhs.node;
+}
+
+template <typename Key, typename Value>
+template <bool IsConst>
+bool
+radix_tree<Key, Value>::radix_tree_iterator<IsConst>::operator==(
+	const radix_tree_iterator<IsConst> &rhs) const
+{
+	return !(*this != rhs);
+}
+
+/*
+ * Returns next leaf (either with smaller or larger key, depending on
+ * ChildIterator type). This function might need to traverse the
+ * tree upwards.
+ */
+template <typename Key, typename Value>
+template <typename ChildIterator>
+typename radix_tree<Key, Value>::tagged_node_ptr const *
+radix_tree<Key, Value>::next_leaf(ChildIterator child_slot)
+{
+	auto parent = child_slot.get_node();
+
+	do {
+		++child_slot;
```

Comment:
So you intentionally go out of bounds and check if the new pointer is outside of the end in parent?:D
Seems a bit hacky to me.